### ISSUE TITLE
MyPy doesn't understand Module

### ISSUE THREAD

**Issue by NeilGirdhar** [patrick-kidger/equinox#820]
MyPy only looks at the first definition of Module, so it doesn't see the redefinition in `TYPE_CHECKING`:
```python
import dataclasses
from typing import TYPE_CHECKING, dataclass_transform


class _ModuleMeta(type):  # This needs to be in a if not TYPE_CHECKING block.
    pass


if TYPE_CHECKING:
    @dataclass_transform(field_specifiers=(dataclasses.field,))
    class _ModuleMeta(type):
        pass


class Module(metaclass=_ModuleMeta):
    pass


class A(Module):
    x: int


A(1)  # MyPy doesn't realize that A is a dataclass
```

**Comment by patrick-kidger** [comment]
Ah -- this must be a mypy vs pyright difference.

Happy to take a PR on this :)

**Comment by patrick-kidger** [comment]
Fixed in #822 :)
